packages/babel-plugin-transform-classes/src/transformClass.js

```diff
@@ -496,10 +497,16 @@ export default function transformClass(
     method: { type: "ClassMethod" },
     path: NodePath,
   ) {
-    // https://github.com/babel/babel/issues/1077
-    if (path.scope.hasOwnBinding(classState.classRef.name)) {
-      path.scope.rename(classState.classRef.name);
-    }
+    const replaceSupers = new ReplaceSupers({
+      methodPath: path,
+      objectRef: classState.classRef,
+      superRef: classState.superName,
+      isLoose: classState.isLoose,
+      file: classState.file,
+      refToPreserve: classState.classRef,
+    });
+
+    replaceSupers.replace();
```

Comment:
Isn't the `ReplaceSupers` in `pushBody` enough?

Reply:
Nope. `classState.classRef` may be shadowed in child scopes of constructor.  You can view the output of `name-collisions-with-class-ref-in-constructor/input.js` applied by the latest babel.
https://babeljs.io/repl#?browsers=%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBiD29oFMAeAXZA7AJjAQpMtAN4BQ00w8WE6ATgK7Drz0AUAlKRZdAJYAzaOwaNk3cnz4hk6OIgDcvaREYAHZPQB0AWzkALeDi7LpAX16Xz0MkA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.12.12&externalPlugins=

```js
  function Foo() {
    var _thisSuper, _this;
    _classCallCheck(this, Foo);

    if (true) {
      var _Foo;
      // `_Foo.prototype` should be `Foo.prototype`
      _get((_thisSuper = _assertThisInitialized(_this), _getPrototypeOf(_Foo.prototype)), "method", _thisSuper).call(_thisSuper);
    }

    return _possibleConstructorReturn(_this);
  }
```

Reply:
Could you add a test that fails without those lines then? I tried commenting them out but `yarn jest classes` still passes.

Reply:
🤔 You are right, the constructor is already transformed in `pushBody`.